src/EFCore.Relational/Metadata/Conventions/SharedTableConvention.cs

```
@@ -584,4 +596,74 @@ protected virtual bool CheckConstraintsUniqueAcrossTables
 
         return null;
     }
+
+    private void TryUniquifyTriggerNames(
+        IConventionEntityType entityType,
+        Dictionary<string, IConventionTrigger> triggers,
+        in StoreObjectIdentifier storeObject,
+        int maxLength)
+    {
+        foreach (var trigger in entityType.GetDeclaredTriggers())
+        {
+            var triggerName = trigger.GetName(storeObject);
+            if (triggerName == null)
+            {
+                continue;
+            }
+
+            if (!triggers.TryGetValue(triggerName, out var otherTrigger))
+            {
+                triggers[triggerName] = trigger;
+                continue;
+            }
+
+            if (AreCompatible(trigger, otherTrigger, storeObject))
+            {
+                continue;
+            }
+
+            var newConstraintName = TryUniquify(trigger, triggerName, triggers, maxLength);
+            if (newConstraintName != null)
+            {
+                triggers[newConstraintName] = trigger;
+                continue;
+            }
+
+            var newOtherTrigger = TryUniquify(otherTrigger, triggerName, triggers, maxLength);
+            if (newOtherTrigger != null)
+            {
+                triggers[triggerName] = trigger;
+                triggers[newOtherTrigger] = otherTrigger;
+            }
+        }
+    }
+
+    /// <summary>
+    ///     Gets a value indicating whether two triggers with the same name are compatible.
+    /// </summary>
+    /// <param name="trigger">A trigger.</param>
+    /// <param name="duplicateTrigger">Another trigger.</param>
+    /// <param name="storeObject">The identifier of the store object.</param>
+    /// <returns><see langword="true" /> if compatible</returns>
+    protected virtual bool AreCompatible(
+        IReadOnlyTrigger trigger,
+        IReadOnlyTrigger duplicateTrigger,
+        in StoreObjectIdentifier storeObject)
+        => false;
```

Should return `true` as currently there's nothing conflicting that could be specified for triggers

---

I thought about it the other way - when providers implement triggers, if they don't customize here it seems better to throw by default (so false) rather than assume compatibility... Especially since any config divergence would make the triggers incompatible anyway. Let me know if you prefer me to change.

---

Yes, but when we start adding the triggers by convention there will likely to be duplicates. If we want to make sure that the providers override this it's better to add a specification test that expects this to throw by default, but can be overridden to not throw.

---

Makes sense, will do.

---

Changed default to true. Thinking about it again, not sure it makes sense to add a test. For any other metadata thing (e.g. index), if the provider adds their own annotations it's their responsibility to override and ensure the compatibility; triggers don't seem any different.